File: daily/Binary Search/744.Find-smallest-letter-greater-than-target.go
package daily

func nextGreatestLetter(letters []byte, target byte) byte {
	var ans byte
	low, high := 0, len(letters)-1
	for low <= high {
		mid := (high + low) / 2
		if letters[mid] > target {
			ans = letters[mid]
			high = mid - 1
		} else {
			low = mid + 1
		}
	}

	if ans != 0 {
		return ans
	}
	return letters[0]
}

File: daily/Binary Search/744.Find-smallest-letter-greater-than-target.py
from typing import List


class Solution:
    def nextGreatestLetter(self, letters: List[str], target: str) -> str:
        left, right = 0, len(letters) - 1
        candidate = letters[0]
        found = False

        while left <= right:
            mid = (left + right) // 2
            if letters[mid] > target:
                candidate = letters[mid]
                right = mid - 1
                found = True
            else:
                left = mid + 1

        if found:
            return candidate
        return letters[0]
